README
NAME
    Map::Metro - Public transport graphing

SYNOPSIS
        # Install a map
        $ cpanm Map::Metro::Plugin::Map::Stockholm

        # And then
        my $graph = Map::Metro->new('Stockholm', hooks => ['PrettyPrinter'])->parse;

        my $routing = $graph->routing_for('Universitetet', 'Kista');

        # or in a terminal
        $ map-metro.pl route Stockholm Universitetet Kista

    prints

        From Universitetet to Kista
        ===========================

        -- Route 1 (cost 15) ----------
        [   T14 ] Universitetet
        [   T14 ] Tekniska högskolan
        [   T14 ] Stadion
        [   T14 ] Östermalmstorg
        [   T14 ] T-Centralen
        [ * T11 ] T-Centralen
        [   T11 ] Rådhuset
        [   T11 ] Fridhemsplan
        [   T11 ] Stadshagen
        [   T11 ] Västra skogen
        [   T11 ] Solna centrum
        [   T11 ] Näckrosen
        [   T11 ] Hallonbergen
        [   T11 ] Kista

        T11  Blue line
        T14  Red line

        *: Transfer to other line
        +: Transfer to other station

DESCRIPTION
    The purpose of this distribution is to find the shortest unique
    route/routes between two stations in a transport network.

    See Task::MapMetro::Maps for a list of released maps.

  Methods
   new($city, hooks => [])
    $city

    The name of the city you want to search connections in. Mandatory,
    unless you are only going to call "available_maps".

    $hooks

    Array reference of Hooks that listens for events.

   parse()
    Returns a Map::Metro::Graph object containing the entire graph.

   available_maps()
    Returns an array reference containing the names of all Map::Metro maps
    installed on the system.

  What is a unique path?
    The following rules are a guideline:

    If the starting station and finishing station...

    ...is on the same line there will be no transfers to other lines.

    ...shares multiple lines (e.g., both stations are on both line 2 and 4),
    each line constitutes a route.

    ...are on different lines a transfer will take place at a shared
    station. No matter how many shared stations there are, there will only
    be one route returned (but which transfer station is used can differ
    between queries).

    ...has no shared stations, the shortest route/routes will be returned.

MORE INFORMATION
    Map::Metro::Graph - How to use graph object.

    Map::Metro::Plugin::Map - How to make your own maps.

    Map::Metro::Hook - How to extend Map::Metro via hooks/events.

    Map::Metro::Cmd - A guide to the command line application.

    Map::Metro::Graph::Connection - Defines a MMG::Connection.

    Map::Metro::Graph::Line - Defines a MMG::Line.

    Map::Metro::Graph::LineStation - Defines a MMG::LineStation.

    Map::Metro::Graph::Route - Defines a MMG::Route.

    Map::Metro::Graph::Routing - Defines a MMG::Routing.

    Map::Metro::Graph::Segment - Defines a MMG::Segment.

    Map::Metro::Graph::Station - Defines a MMG::Station.

    Map::Metro::Graph::Step - Defines a MMG::Step.

    Map::Metro::Graph::Transfer - Defines a MMG::Transfer.

  Hierarchy
    The following is a conceptual overview of the various parts of a graph:

    At first, the map file is parsed. The four types of blocks (stations,
    transfers, lines and segments) are translated into their respective
    object.

    Next, lines and stations are put together into LineStations. Every two
    adjacent LineStations are put into two Connections (one for each
    direction).

    Now the network is complete, and it is time to start traversing it.

    Once a request to search for paths between two stations is given, we
    first search for the starting Station given either a station id or
    station name. Then we find all LineStations for that station.

    Then we do the same for the destination station.

    And then we walk through the network, from LineStation to LineStation,
    finding their Connections and turning them into Steps, which we then add
    to the Route.

    All Routes between the two Stations are then put into a Routing, which
    is returned to the user.

PERFORMANCE
    Since 0.2100 performance is less than an issue than it used to be, but
    it can still be improved. Prior to this version the entire network was
    analyzed up-front. This is unnecessary when searching one (or a few)
    routes. For long-running applications it is still possible to
    pre-calculate all paths, see asps.

STATUS
    This is somewhat experimental. I don't expect that the map file format
    will *break*, but it might be extended. Only the documented api should
    be relied on, though breaking changes might occur.

    For all maps in the Map::Metro::Plugin::Map namespace (unless noted):

    * These maps are not an official source. Use accordingly.

    * There should be a note regarding what routes the map covers.

COMPATIBILITY
    Currently only Perl 5.20+ is supported.

Map::Metro or Map::Tube?
    Map::Tube is the main alternative to "Map::Metro". They both have their
    strong and weak points.

    * Map::Tube is faster.

    * Map::Tube is more stable: It has been on Cpan for a long time, and is
    under active development.

    * Map::Metro has (in my opinion) a better map format.

    * Map::Metro supports eg. transfers between stations.

    * See Task::MapMetro::Maps and Task::Map::Tube for available maps.

    * It is possible to convert Map::Metro maps into Map::Tube maps using
    map-metro.pl.

SEE ALSO
    Map::Tube

AUTHOR
    Erik Carlsson <info@code301.com>

COPYRIGHT
    Copyright 2014 - Erik Carlsson

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for Map-Metro

0.2200 2015-01-08T01:10:28Z CET
 [Enhancements]
 - 0.2101 might have not been as fast as advertised. This should be fixed
   now.
 - Since the dump/hoist and serealize/deserealize commands now are slower
   than the route command, those four commands have been removed (and all
   code related to them in other classes).

 [Meta]
 - Adapted to DZP::OurPkgVersion and new Podweaver settings.

0.2101 2015-01-05T14:51:40Z CET
 [Enhancements]
 - Stop using signatures.

0.2100 2015-01-05T14:43:04Z CET
 [Documentation]
 - Added a short comparison between Map::Metro and Map::Tube.

 [Enhancements]
 - By default, paths are now only searched between the chosen stations.
   This leads to significantly improved performance when searching one/few
   routes per execution.
 - Remoted use of 'aliased', for an extra (and easy) performance gain.
 - Can override cost of line changes (to force route finder to follow
   lines). Undocumented.

0.2002 2015-01-04T13:12:02Z CET
 [Meta]
 - Added prereq (was missing since the map api change in 0.1900)

0.2001 2015-01-03T11:48:46Z CET
 [Enhancements]
 - Added width option on lines.
 - MM::Cmd::Graphviz takes an optional string parameter to set custom
   lengths on edges. The string can alternatively be a path to a file
   containing the settings. This feature is undocumented.

0.2000 2015-01-02T22:51:13Z CET
 [Bug Fixes]
 - Comments (in the map file) must be at the beginning of a line, or
   preceeded only by whitespace.

 [Enhancements]
 - Lines can have options. So far only 'color'.

 [New Features]
 - Introduced graphviz command, which creates a schematic view of a
   network.

0.1901 2015-01-01T01:01:42Z CET
 [Documentation]
 - Mention Dist::Zilla::MintingProfile::MapMetro::Map

0.1900 2014-12-30T22:16:46Z CET
 [API Changes]
 - The MM::Plugin::Map api is changed. The old api will continue to work
   for now (this is not a deprecation notice). The new api allows
   MM::Cmd::Serealize to know the version of the map.

 [Bug Fixes]
 - Removed next_line_station/previous_line_station from LineStation, and
   the setting of them in MM::Graph. These weren't used, and made it harder
   to map a network (eg Oslo) where a line stops at a station in one
   direction, but not in the other.

 [Enhancements]
 - MM::Cmd::Serealize now saves the Sereal-ized file in the
   File::ShareDir::dist_dir of the map distribution.
 - MM::Cmd::Deserealize can now also (for maps serealized starting with
   this version) deserealize maps given the city name (previously you could
   only give a file path).
 - If there is a Serealized file for the current map (and, if applicable,
   version), *and* no hooks are attached, then that file will be
   deserealized and used (instead of re-building the network).

0.1804 2014-12-19T19:17:48Z CET
 [Meta]
 - Add github as issue tracker/repo

0.1803 2014-12-15T14:15:34Z CET
 [Documentation]
 - Added a note on the performance issue on large maps.

 [Enhancements]
 - Added serealize and deserealize commands to MM::Cmd. Helps a bit with
   the performance issue.
 - By default, station names with diacritics (like å, ä and ö) are added
   with their diacritics removed as a search name. This can be turned off
   if it causes problems. See MM::Plugin::Map.

0.1802 2014-12-13T19:13:25Z CET
 [Bug Fixes]
 - Line ids can again be both letters and digits.

 [Documentation]
 - Fixed error in synopsis.

0.1801 2014-12-13T17:19:37Z CET
 [Bug Fixes]
 - available_maps() and map-metro.pl available works again.
 - MM::Emitter should use MM::Hook.

 [Enhancements]
 - Added notation to add extra names (for search and display) to stations.

0.1800 2014-12-13T15:12:14Z CET
 [API Changes]
 - Changed notation to describe segments where a line travels in one
   direction.

0.1702 2014-12-13T12:52:32Z CET
 [Bug Fixes]
 - Thanks to SREZIC:
 - Added missing prereqs
 - Fixed shebang in map-metro.pl

 [Meta]
 - Removed MooseX::Object::Pluggable prereq (replaced by Module::Pluggable
   in 0.1700)

0.1701 2014-12-13T02:04:19Z CET
 [API Changes]
 - Removed a couple of undocumented methods from LineStation. Their purpose
   is better handled by Step.

 [Bug Fixes]
 - Suppressed much warnings during build.

 [Enhancements]
 - Added one-way option to segments.
 - Rewrote Lines command to step through lines using Step (instead of
   LineStation).

0.1700 2014-12-12T23:54:30Z CET
 [API Changes]
 - Renamed routing_completed hook to before_add_routing.
 - Revamped plugin systems for maps.

 [Bug Fixes]
 - In the output from MM::Cmd::Station the last line wouldn't be printed.

 [Documentation]
 - Fleshed out hook example.

 [Enhancements]
 - Argument hooks to Map::Metro->new can now be a string (when only using
   one hook).
 - Added before_add_station hook.
 - Added original_name to Station, so if before_add_station changes the
   station's name, we can still search for the old name (esp when creating
   Segments, but also in routing_for()).
 - Added MM::Plugin::Hook::StreamStations, mostly as a hook example.

0.1600 2014-12-12T02:29:55Z CET
 [API Changes]
 - Removed add to_text() (and related) methods from MM::Graph::*.

 [Enhancements]
 - All relevant MM::Cmd classes are adapted to the new hook system.
 - Ensures that maps only resides in the MM::Plugin::Map namespace.

 [New Features]
 - Created an event/hook system (so far only emits 'routing_completed').
 - Implemented a MM::Plugin::Hook::PrettyPrinter that auto-prints on
   routing_completed.

0.1501 2014-12-11T15:39:01Z CET
 [Enhancements]
 - Added dump and hoist commands to map-metro.pl

0.1500 2014-12-10T19:06:42Z CET
 [Meta]
 - More prereqs.

0.1499 2014-12-10T18:09:34Z CET
 [Meta]
 - Missing prereq.

0.1498 2014-12-10T17:23:49Z CET
 [Meta]
 - Missing prereq.

0.1497 2014-12-10T15:09:17Z CET
 [Meta]
 - Added test prereq

0.1496 2014-12-10T14:24:35Z CET
 [Documentation]
 - Typo

0.1495 2014-12-10T13:12:02Z CET
 [API Changes]
 - Renamed get_route() in MM::Graph to routing_for().

 [Documentation]

lib/Map/Metro.pm
use 5.20.0;
use Map::Metro::Standard;

our $VERSION = '0.2200'; # VERSION
# ABSTRACT: Public transport graphing

package Map::Metro {

    use Moose;
    use Module::Pluggable search_path => ['Map::Metro::Plugin::Map'], require => 1, sub_name => 'system_maps';
    use MooseX::AttributeShortcuts;
    use Types::Standard -types;
    use Types::Path::Tiny -types;
    use List::AllUtils 'any';
    use experimental qw/postderef/;

    use Map::Metro::Graph;

    has map => (
        is => 'ro',
        traits => ['Array'],
        isa => ArrayRef,
        predicate => 1,
        handles => {
            get_map => 'get',
        },
    );
    has mapclasses => (
        is => 'ro',
        traits => ['Array'],
        isa => ArrayRef,
        default => sub { [] },
        handles => {
            add_mapclass => 'push',
            get_mapclass => 'get',
        },
    );
    has hooks => (
        is => 'ro',
        isa => ArrayRef[ Str ],
        traits => ['Array'],
        default => sub { [] },
        handles => {
            all_hooks => 'elements',
            hook_count => 'count',
        },
    );
    has _plugin_ns => (
        is => 'ro',
        isa => Str,
        default => 'Plugin::Map',
        init_arg => undef,
    );

    around BUILDARGS => sub {
        my ($orig, $class, @args) = @_;
        my %args;
        if(scalar @args == 1) {
            $args{'map'} = shift @args;
        }
        elsif(scalar @args % 2 != 0) {
            my $map = shift @args;
            %args = @args;
            $args{'map'} = $map;
        }
        else {
            %args = @args;
        }

        if(exists $args{'map'} && !ArrayRef->check($args{'map'})) {
            $args{'map'} = [$args{'map'}];
        }
        if(exists $args{'hooks'} && !ArrayRef->check($args{'hooks'})) {
            $args{'hooks'} = [$args{'hooks'}];
        }

        return $class->$orig(%args);
    };

    sub BUILD {
        my $self = shift;
        my @args = @_;

        if($self->has_map) {
            my @system_maps = map { s{^Map::Metro::Plugin::Map::}{}; $_ } $self->system_maps;
            if(any { $_ eq $self->get_map(0) } @system_maps) {
                my $mapclass = 'Map::Metro::Plugin::Map::'.$self->get_map(0);
                my $mapobj = $mapclass->new(hooks => $self->hooks);
                $self->add_mapclass($mapobj);
            }
        }
    }

    # Borrowed from Mojo::Util
    sub decamelize {
        my $self = shift;
        my $string = shift;

        return $string if $string !~ m{[A-Z]};
        return join '_' => map {
                                  join ('_' => map { lc } grep { length } split m{([A-Z]{1}[^A-Z]*)})
                               } split '::' => $string;
    }

    sub parse {
        my $self = shift;
        my %args = @_;

        return Map::Metro::Graph->new(filepath => $self->get_mapclass(0)->maplocation,
                                      do_undiacritic => $self->get_mapclass(0)->do_undiacritic,
                                      wanted_hook_plugins => [$self->all_hooks],
                                      exists $args{'override_line_change_weight'} ? (override_line_change_weight => $args{'override_line_change_weight'}) : (),
                                )->parse;
    }

    sub available_maps {
        my $self = shift;
        return sort $self->system_maps;
    }
}

__END__

=pod

=encoding utf-8

=head1 NAME

Map::Metro - Public transport graphing

=head1 VERSION

Version 0.2200, released 2015-01-08.

=head1 SYNOPSIS

    # Install a map
    $ cpanm Map::Metro::Plugin::Map::Stockholm

    # And then
    my $graph = Map::Metro->new('Stockholm', hooks => ['PrettyPrinter'])->parse;

    my $routing = $graph->routing_for('Universitetet', 'Kista');

    # or in a terminal
    $ map-metro.pl route Stockholm Universitetet Kista

prints

    From Universitetet to Kista
    ===========================

    -- Route 1 (cost 15) ----------
    [   T14 ] Universitetet
    [   T14 ] Tekniska högskolan
    [   T14 ] Stadion
    [   T14 ] Östermalmstorg
    [   T14 ] T-Centralen
    [ * T11 ] T-Centralen
    [   T11 ] Rådhuset
    [   T11 ] Fridhemsplan
    [   T11 ] Stadshagen
    [   T11 ] Västra skogen
    [   T11 ] Solna centrum
    [   T11 ] Näckrosen
    [   T11 ] Hallonbergen
    [   T11 ] Kista

    T11  Blue line
    T14  Red line

    *: Transfer to other line
    +: Transfer to other station

=head1 DESCRIPTION

The purpose of this distribution is to find the shortest L<unique|/"What is a unique path?"> route/routes between two stations in a transport network.

See L<Task::MapMetro::Maps> for a list of released maps.

=head2 Methods

=head3 new($city, hooks => [])

B<C<$city>>

The name of the city you want to search connections in. Mandatory, unless you are only going to call L</"available_maps">.

B<C<$hooks>>

Array reference of L<Hooks|Map::Metro::Hook> that listens for events.

=head3 parse()

Returns a L<Map::Metro::Graph> object containing the entire graph.

=head3 available_maps()

Returns an array reference containing the names of all Map::Metro maps installed on the system.

=head2 What is a unique path?

The following rules are a guideline:

If the starting station and finishing station...

...is on the same line there will be no transfers to other lines.

...shares multiple lines (e.g., both stations are on both line 2 and 4), each line constitutes a route.

...are on different lines a transfer will take place at a shared station. No matter how many shared stations there are, there will only be one route returned (but which transfer station is used can differ between queries).

...has no shared stations, the shortest route/routes will be returned.

=head1 MORE INFORMATION

L<Map::Metro::Graph> - How to use graph object.

L<Map::Metro::Plugin::Map> - How to make your own maps.

L<Map::Metro::Hook> - How to extend Map::Metro via hooks/events.

L<Map::Metro::Cmd> - A guide to the command line application.

L<Map::Metro::Graph::Connection> - Defines a MMG::Connection.

L<Map::Metro::Graph::Line> - Defines a MMG::Line.

L<Map::Metro::Graph::LineStation> - Defines a MMG::LineStation.

L<Map::Metro::Graph::Route> - Defines a MMG::Route.

L<Map::Metro::Graph::Routing> - Defines a MMG::Routing.

L<Map::Metro::Graph::Segment> - Defines a MMG::Segment.

L<Map::Metro::Graph::Station> - Defines a MMG::Station.

L<Map::Metro::Graph::Step> - Defines a MMG::Step.

L<Map::Metro::Graph::Transfer> - Defines a MMG::Transfer.

=head2 Hierarchy

The following is a conceptual overview of the various parts of a graph:

At first, the map file is parsed. The four types of blocks (stations, transfers, lines and segments) are translated
into their respective object.

Next, lines and stations are put together into L<LineStations|Map::Metro::Graph::LineStation>. Every two adjacent LineStations
are put into two L<Connections|Map::Metro::Graph::Connection> (one for each direction).

Now the network is complete, and it is time to start traversing it.

Once a request to search for paths between two stations is given, we first search for the starting L<Station|Map::Metro::Graph::Station> given either a
station id or station name. Then we find all L<LineStations|Map::Metro::Graph::LineStation> for that station.

Then we do the same for the destination station.

And then we walk through the network, from L<LineStation|Map::Metro::Graph::LineStation> to L<LineStation|Map::Metro::Graph::LineStation>, finding their L<Connections|Map::Metro::Graph::Connection>
and turning them into L<Steps|Map::Metro::Graph::Step>, which we then add to the L<Route|Map::Metro::Graph::Route>.

All L<Routes|Map::Metro::Graph::Route> between the two L<Stations|Map::Metro::Graph::Station> are then put into a L<Routing|Map::Metro::Graph::Routing>, which is returned to the user.

=head1 PERFORMANCE

Since 0.2200 performance is less than an issue than it used to be, but it can still be improved. Prior to this version the entire network was analyzed up-front. This is unnecessary when searching one (or a few) routes. For long-running applications it is still possible to pre-calculate all paths, see L<asps|Map::Metro::Graph/"asps()">.

=head1 STATUS

This is somewhat experimental. I don't expect that the map file format will I<break>, but it might be
extended. Only the documented api should be relied on, though breaking changes might occur.

For all maps in the Map::Metro::Plugin::Map namespace (unless noted):

* These maps are not an official source. Use accordingly.

* There should be a note regarding what routes the map covers.

=head1 COMPATIBILITY

Currently only Perl 5.20+ is supported.

=head1 Map::Metro or Map::Tube?

L<Map::Tube> is the main alternative to C<Map::Metro>. They both have their strong and weak points.

* Map::Tube is faster.

* Map::Tube is more stable: It has been on Cpan for a long time, and is under active development.

* Map::Metro has (in my opinion) a better map format.

* Map::Metro supports eg. transfers between stations.

* See L<Task::MapMetro::Maps> and L<Task::Map::Tube> for available maps.

* It is possible to convert Map::Metro maps into Map::Tube maps using L<map-metro.pl|Map::Metro::Cmd/"map-metro.pl metro_to_tube $city">.

=head1 SEE ALSO

L<Map::Tube>

=head1 SOURCE

L<https://github.com/Csson/p5-Map-Metro>

=head1 HOMEPAGE

L<https://metacpan.org/release/Map-Metro>

=head1 AUTHOR

Erik Carlsson <info@code301.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2015 by Erik Carlsson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
